src/main/java/com/artipie/asto/fs/FileStorage.java

```diff
@@ -72,27 +74,24 @@ public FileStorage(final Path path) {
     }
 
     @Override
-    public CompletableFuture<Collection<Key>> list(final String prefix) {
+    public CompletableFuture<Collection<Key>> list(final Key prefix) {
         return Single.fromCallable(
             () -> {
-                final String separator = FileSystems.getDefault().getSeparator();
-                if (!prefix.endsWith(separator)) {
-                    throw new IllegalArgumentException(
-                        String.format(
-                            "The prefix must end with '%s': \"%s\"",
-                            prefix,
-                            separator
-                        )
-                    );
+                final Path path = this.path(prefix);
+                final Collection<Key> keys;
+                if (Files.exists(path)) {
+                    final int dirnamelen = path.toString().length() - prefix.string().length();
+                    keys = Files.walk(path)
+                        .filter(Files::isRegularFile)
+                        .map(Path::toString)
+                        .map(p -> p.substring(dirnamelen))
+                        .map(s -> s.split(FileSystems.getDefault().getSeparator()))
```

Comment:
@olegmoz seems that now the tests could fail on the wrong filesystem (on Windows, for example). Can we prevent this somehow? Or write system-dependent tests (but I don't like this idea). It's better to write system independent tests.

Reply:
@iakunin on the contrary, before my last commits it failed on Unix. Now tests run on both Unix and Windows, see the checks - we run tests on both systems.

Reply:
@olegmoz I see. Now it's much clearer for me, why it was done. Thanks for your explanation.